### ISSUE TITLE
CTRL, ALT, TAB, ESC keys addition

### ISSUE THREAD

**Issue by FraYoshi** [futo-org/android-keyboard#25]
These keys are a very useful, especially for Emacs users, and would be great to have them as a long press key, dedicated, or even also like now we have the arrows, redo/undo, copy/paste.. in a dedicated row. In such case, I would prefer it not as big as it currently is.. maybe just the height of a letters row ^^

Attached 1 screenshot with the example of Unexpected Keyboard, where you slide to one angle, and you get 'Alt', 'Esc' and 'Tab', while the  'Ctrl' is a dedicated button.

![FUTO-ctrl-alt-tab-esc-Unexpected](https://github.com/futo-org/android-keyboard/assets/14005514/1554e716-1d78-488f-b338-5886ad901432)

**Comment by domportera** [comment]
yeah i wont be able to leave Unexpected Keyboard completely without this. since theres a PC QW|RTY layout, i would have expected it to be there. Ctrl Z and Ctrl Y are super handy over opening the additional menu, and its also really great for remoting into my PC and advanced text editing

**Comment by favdish** [comment]
I am in the same boat. Much Terminal work I do just needs Ctrl, Tab and Esc, so can't leave my current keyboard as long as these are not available.

**Comment by Solar-Simon** [comment]
'TAB' is also necessary to indent points in lists in some editors.

**Comment by Julrob199** [comment]
i am really hoping to have this implemented 🙏

**Comment by chaosORCHESTRA** [comment]
> Hi, I'm trying to find a way to add the Tab key to my keyboard. Has this been implemented?


Me too. Even having it in the action bar would be useful

**Comment by lepras** [comment]
Termux has this layout:

![Image](https://github.com/user-attachments/assets/b7c66e7d-3aa7-44ab-9c16-eeb578124e5e)

We can have that here too, if you think it will clutter the keyboard for "non-power" users you can have an option in futo keyboard to enable it here:

Keyboard & typing > Advance > enable ctrl, alt, tab.

**Comment by Julrob199** [comment]
> We can have that here too, if you think it will clutter the keyboard for "non-power" users you can have an option in futo keyboard to enable it here:
> 
> Keyboard & typing > Advance > enable ctrl, alt, tab.

But it will take "precious" vertical space.

Would it be possible to add a button and add a popup on the shift key or some other keys like comma or point (like in OP's post the screenshots)? or put the buttons on a third page next to the special characters?

**Comment by haldi4803** [comment]
Would be great to get Tab/Enter/Esc/Ctrl as Actions for now.

So everyone who needs them can have then in the action bar and those who don't want to don't waste space.

**Comment by Julrob199** [comment]
> Would be great to get Tab/Enter/Esc/Ctrl as Actions for now.

If we come up with a solution, it should be future proof and not a temporary fix, as called "for now".

Like mentioned above, the action bar is not intended for regular keys imo but rather for functions (paste, emojis, ...). A more sustainable approach might be to introduce key mapping or a pop-up function for these inputs instead. But whatever solution we go for, it should be a sustainable and long term solution. What do you think?

What options do we have?
- Put them in the action bar like @haldi4803 said
- Create a new "mode" for the keyboard, with arrow keys (they are already in the action bar I know) and esc, ctrl, alt, tab, ...
- Create some pop up (long term click)

What is technically the easiest solution and what is the technically correct, best user experience based solution?

**Comment by abb128** [comment]
From my understanding there technically wouldn't really be an easy or correct way to implement this while still supporting autocorrect and text composition functionality, so most likely it would have to be a separate layout option that has that functionality turned off and behaves more similarly to Hacker's Keyboard or Unexpected Keyboard when it's selected

**Comment by Zvonimir-FUTO** [comment]
from #1260

### Description

There used to be a programmers keyboard with features useful in text editing. It would be better to have a scrollable keyboard to get more screen estate than to have all these switches between keyboards. This would allow for not only a standard PC keyboard but an extended one. Especially useful when using terminal emulators, computer emulators and remote desktop

    Arrow buttons
    Home, end etc.
    Function keys. F1, F2 etc
    Sun Unix buttons. Copy, paste etc
    Ctrl, alt, meta, etc
    Extra buttons. Audio volume, screen brightness etc
    On/Off an reboot. Highly useful together with emulators but. Also useful for the phone itself.

My favorite keyboard of all time was a sun type 6/7. So that's a good base for this. It has all the standard PC keys an then add some.

### Complexity

Using scrolling and having a single keyboard is probably easier to implement than a standard Google keyboard with all its subkeyboards. Add zooming and it will adapt to display and finger size.

Instead of being mindful to fit keys on screen there is unlimited space.

**Comment by matthew-hoad** [comment]
> Like mentioned above, the action bar is not intended for regular keys imo but rather for functions (paste, emojis, ...).

I think given the fact we have arrow keys in the actions menu, and they can be pinned under the keyboard also, treating the additional power user keys the same way would be great. Even if just in the action menu. As mentioned in the thread, it gives power users the option to include them, even pin them, but it's just a customisation option.

Alternatively there is the way the "Text Editor" works, is something similar possible here where you can open it temporarily, and optionally expand the keyboard to show underneath it? Honestly if I could customize what keys are in a lanel like that, it would be pretty sweet.

**Comment by DrTexx** [comment]
+1 I would also really love to see this, especially the ability to send tab / shift-tab input to indent/unindent in apps like Logseq or Obsidian :)

I too wouldn't mind it in the action bar thing, that's where I pinned the up and down keys recently and it's been absolutely excellent 💚 :>

**Comment by kuba2008a2-n0** [comment]
Where on earth is the ```TAB``` key?
When will it finally be implemented?
Its a really crucial feature.

**Comment by anthonywilliams** [comment]
Please add this feature, even if I have to use a custom layout to get the keys. As others have said, I like FUTO as my main keyboard, and it is frustrating to have to switch to Unexpected keyboard for emacs and terminal applications.

**Comment by MN-Core** [comment]
I agree that this would be a nice feature. I personally, could use the Tab key for the sake of indenting.

Implementing it as an acrion that can be used in the action bar is one route but as others have said, this takes up precious vertical space.

---

So I have a couple of suggestions for implementation:


### Added as an extra key when holding down a letter

Currently, holding `A` allows you to `Select All` and holding `Z` allows you to `Undo` much like holding Ctrl+A and Ctrl+Z on physical keyboard.

However, the choice of letters for these shortcuts might make this keyboard more opinionated.


### Alternatively, an extra menu using a short/long press and swipe.

Perhaps a swipe and hold interaction from the space to the left of the first letter on the second row (the `a` button on a qwerty layout), could open a new control.
This might be a bar or a pie wheel with options, users could then release the press on an option like Tab, Reverse tab, esc etc.

### Lastly, a triple tap to open a dev keyboard layout

Triple tapping the shift key could clear the keyboard space of all letters and replace with commands for power users and formatting.
Similar to how pressing `?123` opens a Numbers and Symbols keyboard layout. The user would be locked in this keyboard layout unless they hit the highlighted Shift key again.
But then how this overall interaction is executed could also be very opinionated.

---

I'm just a junior dev but I'd be happy to contribute if there is a consensus on how it should be implemented. Thanks for the work you all put into this keyboard, I appreciate it very much!

**Comment by kuba2008a2-n0** [comment]
It should be added In the secondary menu.
**there are literally 2 same `percentage` keys in there**
I think converting one into a `tab` key wouldnt be that bad.

<img width="800" height="688" alt="Image" src="https://github.com/user-attachments/assets/bcc21098-c967-48b0-9797-07f57bb41e39" />

**Comment by Alastair-L** [comment]
This would be great for me! Tab especially, but also home/end as I use all three a lot for text editing.

Additionally these "advanced" keys could be useful in apps that I know have other ways to do things but I just know the desktop way (escape in excel). 

I'd be happy for them to be hidden in a menu or actions

**Comment by Fonant** [comment]
This would be VERY useful for me!

Particularly TAB (command completion), CTRL and ESC.

**Comment by suderman** [comment]
I definitely need esc, tab, ctrl, alt keys a lot more than I need ‰, №, ¢, ♪, §, or most of these other obscure keys provided.

**Comment by chrsdk** [comment]
I use Heliboard when I need special keys like ESC. I added a custom layout in Heliboard so that when I long press "e" for example I get a pop-up which includes ESC beside all the usual ē, ê, etc. It's easy to remember where it is and doesn't cost any extra screen space.

**Comment by madushan1000** [comment]
Maybe make it possible to send custom linux keycodes? As far as I can see, only a per-defined set of key codes are available to use even in custom keyboard layouts.
Android supports a lot of keycodes, you can find a list here: https://developer.android.com/reference/android/view/KeyEvent#constants_1
